### PR TITLE
x-pack/plugin/core: make automatic rollovers lazy

### LINKED ISSUE #104083: Update internal unconditional rollover calls to use lazy rollover
### Description

Currently we have some of our `IndexTemplateRegistry` subclasses that invoke unconditional rollovers after templates have been upgraded. One thing that would be good to do is to make use of the recently merged [lazy rollover flag](https://github.com/elastic/elasticsearch/pull/103309) for these calls.

We can look at uses of the `applyRolloverAfterTemplateV2Upgrade()` flag to see where these are used.

### PR DESCRIPTION
**Pull request by axw**
Apply automatic rollovers lazily, upon the next write, to avoid unnecessary empty backing indices.

See https://github.com/elastic/elasticsearch/issues/89346

Fixes #104083

### CODE REVIEW

**Comment by axw** [comment]
Failure is in ESQL, unrelated to my changes AFAICS.

**Comment by felixbarny** [comment]
@elasticmachine run elasticsearch-ci/part-3

**Review by felixbarny — APPROVED** [review]

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java [hunk 1]
```diff
@@ -814,6 +814,7 @@ public void onFailure(Exception e) {
                         templateName
                     );
                     RolloverRequest request = new RolloverRequest(rolloverTarget, null);
+                    request.lazy(true);
```
[the remark this answers is not in the document]
Comment (axw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:817]: Done

**Comment by axw** [comment]
@elasticmachine update branch

**Review by gmarouli — APPROVED** [review]
Thank you for fixing this @axw. LGTM, if I am not mistaken this is fixing https://github.com/elastic/elasticsearch/issues/104083.

**Comment by axw** [comment]
> Thank you for fixing this @axw. LGTM, if I am not mistaken this is fixing https://github.com/elastic/elasticsearch/issues/104083.

Hah, I didn't see that issue before :) It does indeed. There's just one registry (the APM one) that overrides `applyRolloverAfterTemplateV2Upgrade` and enables auto rollover, and that's desirable for us.

**Comment by axw** [comment]
@elasticmachine update branch

**Pull request merged**